### ISSUE TITLE
Traffic blocked when it should be allowed by netpol

### ISSUE THREAD

**Issue by itsthatriver** [cilium/cilium#20258]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I'm trying to make a connection between pods in two different namespaces. I have removed all network policies in the cluster except for one, which is in the namespace of the destination pod. The network policy is:
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-ingress-to-nginx
  namespace: asc-14997fb1-b878
spec:
  ingress:
  - from:
    - ipBlock:
        cidr: 0.0.0.0/0
  - from:
    - podSelector: {}
  - from:
    - namespaceSelector: {}
  podSelector:
    matchLabels:
      app: nginx-lb
  policyTypes:
  - Ingress
```

We were testing a more restrictive policy but kept running into issues so I tried to work my way back to a fully permissive ingress policy.

From the source pod, I run this (`10.0.9.158` is the IP of the destination pod)):
```
nc -zv 10.0.9.158 443
... time out
```

Full debug details in [Slack thread](https://cilium.slack.com/archives/C1MATJ5U5/p1655764534307359).

### Cilium Version

```
root@ip-192-168-8-134:/home/cilium# cilium version
Client: 1.9.9 5bcf83c 2021-07-19T16:45:00-07:00 go version go1.15.14 linux/amd64
Daemon: 1.9.9 5bcf83c 2021-07-19T16:45:00-07:00 go version go1.15.14 linux/amd64
```

### Kernel Version

```
root@ip-192-168-8-134:/home/cilium# uname -a
Linux ip-192-168-8-134.ec2.internal 5.4.190-107.353.amzn2.x86_64 #1 SMP Wed Apr 27 21:16:35 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

### Kubernetes Version

```
❯ kubectl version
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.12-eks-a64ea69", GitCommit:"d4336843ba36120e9ed1491fddff5f2fec33eb77", GitTreeState:"clean", BuildDate:"2022-05-12T18:29:27Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
```

### Sysdump

[cilium-sysdump-20220620-183155.zip](https://github.com/cilium/cilium/files/8945032/cilium-sysdump-20220620-183155.zip)

### Relevant log output

```shell
# running cilium monitor on destination pod
root@ip-192-168-8-134:/home/cilium# cilium monitor --to 3042
Listening for events on 8 CPUs with 64x4096 of shared memory
Press Ctrl-C to quit
level=info msg="Initializing dissection cache..." subsys=monitor
xx drop (Policy denied) flow 0x31783178 to endpoint 3042, identity 7193->25836: 10.0.32.72:42057 -> 10.0.9.158:443 tcp SYN

# running cilium monitor on source pod
root@ip-192-168-1-26:/home/cilium# cilium monitor --from 2101
Listening for events on 16 CPUs with 64x4096 of shared memory
Press Ctrl-C to quit
level=info msg="Initializing dissection cache..." subsys=monitor
-> overlay flow 0x9d7730db identity 7193->0 state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.0.32.72:43197 -> 10.0.9.158:443 tcp SYN
-> overlay flow 0xf67c6a65 identity 7193->0 state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.0.32.72:49223 -> 10.0.11.98:53 udp
-> overlay flow 0xbd342d34 identity 7193->0 state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.0.32.72:44743 -> 10.0.29.146:53 udp
-> overlay flow 0x81e6dc56 identity 7193->0 state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.0.32.72:55869 -> 10.0.11.98:53 udp

# policy evaluation for the two identities
root@ip-192-168-1-26:/home/cilium# cilium monitor --from 2101 --to 3042
Listening for events on 16 CPUs with 64x4096 of shared memory
Press Ctrl-C to quit
^C
Received an interrupt, disconnecting from monitor...

root@ip-192-168-1-26:/home/cilium# cilium policy trace -s 2101 -d 3042 --dport 443
----------------------------------------------------------------
From: [any:2101] => To: [any:3042] Ports: [443/ANY]
  Policy enforcement is disabled because no rules in the policy repository match any endpoint selector from the provided destination sets of labels.
Label verdict: allowed

Final verdict: ALLOWED
```

### Anything else?

The source is shown as `Unknown App` when traffic is viewed through the hubble ui in the destination namespace
![image (9)](https://user-images.githubusercontent.com/5632093/174698060-ac9524bc-c0ac-45ca-a3cc-876d577391e9.png)
The source is correctly identified when viewed from the source namespace (I did try restarting pods to see if they were unmanaged, and that had no effect)
![image (10)](https://user-images.githubusercontent.com/5632093/174698063-f7e28a09-40fb-41f3-8c73-411280da2c4b.png)

We're running in EKS with VXLAN tunneling and cluster-pool IPAM. We also have restricted the labels for identities of our pods
```
  # To include or exclude matched resources from cilium identity evaluation
  labels: "k8s:io.kubernetes.pod.namespace k8s:k8s-app k8s:app k8s:name k8s:spark-role"
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by thejosephstevens** [comment]
Update, I just tried this CiliumNetworkPolicy as well and still got blocked traffic. I got the code from the docs [here](https://docs.cilium.io/en/stable/policy/kubernetes/#example-expose-pods-across-namespaces)
```
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: allow-ingress-to-nginx
  namespace: asc-14997fb1-b878
spec:
  endpointSelector:
    matchLabels:
      app: nginx-lb
  ingress:
    - fromEndpoints:
        - matchLabels:
            app: kong-dp
            k8s:io.kubernetes.pod.namespace: cell-manager
```

**Comment by kkourt** [comment]
Thank you for the well-written report.

The policy drop seems to happen in the destination node:
> xx drop (Policy denied) flow 0x31783178 to endpoint 3042, identity 7193->25836: 10.0.32.72:42057 -> 10.0.9.158:443 tcp SYN

Unfortunately, the sysdump does not include the destination node bugtool, but only the source (`cilium-bugtool-cilium-btpvb-20220620-183155/` directory) which would be useful to have to investigate the drop. Not sure why this is the case. An alternative would be to use the cilium-bugtool directly to get the information on the destination node: https://docs.cilium.io/en/latest/cmdref/cilium-bugtool/.

**Comment by thejosephstevens** [comment]
Just grabbed one, does this work?
[cilium-bugtool-20220622-004654.396+0000-UTC-887160975.tar.zip](https://github.com/cilium/cilium/files/8953602/cilium-bugtool-20220622-004654.396%2B0000-UTC-887160975.tar.zip)

**Comment by thejosephstevens** [comment]
@kkourt Anything else I can provide that would be helpful? Or anything I should be trying out to address this?

**Comment by kkourt** [comment]
> Just grabbed one, does this work?
[cilium-bugtool-20220622-004654.396+0000-UTC-887160975.tar.zip](https://github.com/cilium/cilium/files/8953602/cilium-bugtool-20220622-004654.396%2B0000-UTC-887160975.tar.zip)

Thanks!

The local identity on the destination pod seems fine:
```
#### Identity get 25836

ID      LABELS
25836   k8s:app=nginx-lb
        k8s:ascend.io/external-service=true
        k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=asc-14997fb1-b878
        k8s:io.kubernetes.pod.namespace=asc-14997fb1-b878
```

The source endpoints do not seem to match the policy, however.

<details>
```
        "policy": {
          "proxy-statistics": [],
          "realized": {
            "allowed-egress-identities": [
              0
            ],
            "allowed-ingress-identities": [
              1
            ],
            "build": 60,
            "cidr-policy": {
              "egress": [],
              "ingress": []
            },
            "denied-egress-identities": null,
            "denied-ingress-identities": null,
            "id": 25836,
            "l4": {
              "egress": [],
              "ingress": [
                {
                  "derived-from-rules": [
                    [
                      "k8s:io.cilium.k8s.policy.derived-from=CiliumNetworkPolicy",
                      "k8s:io.cilium.k8s.policy.name=allow-ingress-to-nginx",
                      "k8s:io.cilium.k8s.policy.namespace=asc-14997fb1-b878",
                      "k8s:io.cilium.k8s.policy.uid=145f6b3f-332d-4938-9093-4f2f1c0b631f"
                    ]
                  ],
                  "rule": "{\n  \"port\": 0,\n  \"protocol\": \"ANY\",\n  \"l7-rules\": [\n    {\n      \"\\u0026LabelSelector{MatchLabels:map[string]string{any.app: kong-dp,k8s.io.kubernetes.pod.namespace: cell-manager,},MatchExpressions:[]LabelSelectorRequirement{},}\": null\n    }\n  ]\n}"
                }
              ]
            },
            "policy-enabled": "ingress",
            "policy-revision": 60
          },
          "spec": {
            "allowed-egress-identities": [
              0
            ],
            "allowed-ingress-identities": [
              1
            ],
            "build": 60,
            "cidr-policy": {
              "egress": [],
              "ingress": []
            },
            "denied-egress-identities": null,
            "denied-ingress-identities": null,
            "id": 25836,
            "l4": {
              "egress": [],
              "ingress": [
                {
                  "derived-from-rules": [
                    [
                      "k8s:io.cilium.k8s.policy.derived-from=CiliumNetworkPolicy",
                      "k8s:io.cilium.k8s.policy.name=allow-ingress-to-nginx",
                      "k8s:io.cilium.k8s.policy.namespace=asc-14997fb1-b878",
                      "k8s:io.cilium.k8s.policy.uid=145f6b3f-332d-4938-9093-4f2f1c0b631f"
                    ]
                  ],
                  "rule": "{\n  \"port\": 0,\n  \"protocol\": \"ANY\",\n  \"l7-rules\": [\n    {\n      \"\\u0026LabelSelector{MatchLabels:map[string]string{any.app: kong-dp,k8s.io.kubernetes.pod.namespace: cell-manager,},MatchExpressions:[]LabelSelectorRequirement{},}\": null\n    }\n  ]\n}"
                }
              ]
            },
            "policy-enabled": "ingress",
            "policy-revision": 60
          }
        },
```


```
/sys/fs/bpf/tc/globals/cilium_policy_03042:

POLICY   DIRECTION   IDENTITY   PORT/PROTO   PROXY PORT   BYTES     PACKETS   
Allow    Ingress     1          ANY          NONE         2442329   13025     
Allow    Egress      0          ANY          NONE         3722893   5968   
```

</details>

which leads to the packet being dropped.

The next thing I would do is log at the agent logs on the receiving node, but unfortunately the agent logs on that node do not exist in the sysdump.

However, one thing that seems suspicious is:

on the destination node:
```
        "name": "sync-to-k8s-ciliumendpoint (3042)",
        "status": {
          "consecutive-failure-count": 440,
          "failure-count": 440,
          "last-failure-msg": "ciliumendpoints.cilium.io \"nginx-lb-dp-5b9679b995-gxfq6\" not found",
          "last-failure-timestamp": "2022-06-22T00:41:38.809Z",
          "last-success-timestamp": "0001-01-01T00:00:00.000Z"
        },
        "uuid": "31ab9236-f0e3-11ec-8b01-12f5b146bf7b"
```

Looking at the source node, the same is true for the other endpoint:
```
        "name": "sync-to-k8s-ciliumendpoint (2101)",
        "status": {
          "consecutive-failure-count": 170,
          "failure-count": 170,
          "last-failure-msg": "ciliumendpoints.cilium.io \"kong-dp-58c577d9b5-4l7hj\" not found",
          "last-failure-timestamp": "2022-06-21T01:29:12.285Z",
          "last-success-timestamp": "0001-01-01T00:00:00.000Z"
        },
        "uuid": "1bad56fc-f0e0-11ec-bbe4-0ec80be517bd"
```

The inability to sync the cilium endpoint information, might explain why the policy does not match any identities on the destination node.

**Comment by thejosephstevens** [comment]
Got it, we had noticed some error logs around failure to sync cilium endpoints but weren't clear on whether they were conclusive or a red herring. What are the typical causes of that we could look at? Logically I think it would make sense as a cause (if the identity is unknown at the destination, a DENY is correct enforcement), I'm mostly surprised that it appears to be non-transient (like we're not seeing this eventually resolve, the failure is totally stable)

**Comment by kkourt** [comment]
Given the "not found" error message,

maybe it's worth digging into the agent logs to check why the endpoint was never created.

**Comment by christarazi** [comment]
This might be related to https://github.com/cilium/cilium/pull/18864. I noticed that the Cilium version is quite out-of-date. The fix mentioned was released in v1.9.14, but that's just one fix the I recall off the top of my head. I'm sure there have been many improvements around CiliumEndpoint lifecycle in the latest version of Cilium. Could you upgrade to the latest?

**Comment by thejosephstevens** [comment]
Yeah, I think we can bump to 1.9.14 no problem (at least to try). We've been trying to orchestrate a jump to 1.11+, but have had issues as we try to get that lined up across all clouds (currently waiting on 1.12 release plus official Azure BYOCNI support before re-attempting). Not sure if I'll get to this today, but should be able to this week

**Comment by thejosephstevens** [comment]
Alright, so I just tried updating to 1.9.17 and it had no impact (I rolled pods after to be sure). What I'm seeing that's different from other working clusters is two things. First, tons of errors like this coming off the agent containers
```
level=error msg="Cannot update CEP" containerID=ea17922727 controller="sync-to-k8s-ciliumendpoint (1921)" datapathPolicyRevision=2 desiredPolicyRevision=2 endpointID=1921 error="ciliumendpoints.cilium.io \"authz-dp-c9b894c98-fdg8w\" not found" identity=12067 ipv4=10.0.51.251 ipv6= k8sPodName=asc-983682d4-8732-48f1/authz-dp-c9b894c98-fdg8w subsys=endpointsynchronizer
```
Second, when I run `kubectl get cep`, every CEP just has a name and namespace, no endpoint ID, Identity, state, or IP.

Based on everything I'm seeing so far, I'm pretty sure these are indicators of the cause, but I'm confused by the error because if I look at the CEP list the actual endpoints the errors reference are definitely there. To make it more confusing, I would expect this to then be something like an auth error, timeout, or something, but the error log itself is giving no indication of that.

Even weirder, if I download kubectl into a cilium agent container and run `kubectl get cep -A`, I get back the full list of cilium endpoints extremely quickly.

I'll keep digging, but this is a head-scratcher for me right now

**Comment by christarazi** [comment]
To be clear, you rolled which pods? Cilium pods or the workload pods?

**Comment by thejosephstevens** [comment]
Workload pods. I actually didn't get all the way through, the command failed partway through the roll referencing too many open connections. Currently I'm trying to recover the cluster because when I attempted to roll nodes as a follow-up, nothing is able to rejoin (unclear why)

**Comment by christarazi** [comment]
And these 

```
level=error msg="Cannot update CEP"
```

errors happened on pods that you initiated a rollout for?

**Comment by thejosephstevens** [comment]
Just got the cluster back up, at this point all nodes and pods have been replaced in the cluster since the bump from 1.9.9 to 1.9.17 and I'm still seeing the same state for CEPs and errors from agents. Every CEP I have spot-checked I have confirmed exists in the cluster. Example:
```
level=error msg="Cannot update CEP" containerID= controller="sync-to-k8s-ciliumendpoint (48)" datapathPolicyRevision=2 desiredPolicyRevision=2 endpointID=48 error="ciliumendpoints.cilium.io \"scheduler-dp-7bc89666d6-zw287\" not found" identity=60409 ipv4= ipv6= k8sPodName=/ subsys=endpointsynchronizer

❯ kubectl get cep -A | grep scheduler-dp-7bc89666d6-zw287
asc-9ea53f0a-06c5           scheduler-dp-7bc89666d6-zw287
```

**Comment by thejosephstevens** [comment]
I'm calling it for the night, but will be working more on this tomorrow. If there's anything you can think of in terms of error types that might be contained in the generic `not found` error that I'm seeing I'd be more than happy to chase down leads once I come back online. Otherwise I'm firing a bit blind. I did take a quick look at the code around this and it's really not that complicated, so not clear what could be going wrong in there.

But in particular, if there's anything like a quiet timeout that might be getting hit that would be good to know.

**Comment by christarazi** [comment]
Can you take a [sysdump](https://docs.cilium.io/en/v1.11/operations/troubleshooting/#automatic-log-state-collection) again which includes a node / cilium-agent which is failing with https://github.com/cilium/cilium/issues/20258#issuecomment-1171995504?

Also, I'd be curious to know if you are continuing to see this issue for newly restarted workloads. My suspicion is that somehow the CEP state was corrupted for unknown reasons and now Cilium cannot recover. I suspect that if you delete the workload pod that is having issues and delete its corresponding CEP (Cilium should do this for you, but ensuring that it's gone would be good), then the new workload shouldn't experience the "Cannot update CEP" error anymore.

**Comment by thejosephstevens** [comment]
Here's one (it's happening on every node constantly). Unfortunately that hasn't produced a fix, we're really familiar with the issues of unmanaged workloads, so our first attempt at a fix was restarting the pods that were having issues, but the issue blockingf CEP update has proved to be completely stable. We just got back from a company break so I'm restarting work on this today. My first debug step is going to be creating a brand new cluster to try and determine at what point CEP updates stop working. We have a few dozen other EKS clusters where cilium netpols work just fine, so my assumption is we changed something in this cluster to break it, but I'm not clear on what yet.
[cilium-sysdump-20220711-120246.zip](https://github.com/cilium/cilium/files/9086876/cilium-sysdump-20220711-120246.zip)

**Comment by thejosephstevens** [comment]
Okay, so I've got to apologize because I'm afraid this is going to remain a mystery. In my investigations earlier I found that the other cluster we had which was identically configured to this one (same config, same workload profiles, everything), was not experiencing the same problems of endpoints updating.

In an attempt to get this issue addressed I redeployed cilium with `clean-cilium-state: "true"` and then rolled nodes to try and force recovery, but found that the same issue still persisted.

Finally, in order to get past this, I replaced the k8s cluster, ran a full deploy or all resources, and I'm not seeing the issue in the new cluster at all.

So apologies that I can't shed more light on the issue that occurred here, because I would love to be able to come out of this issue with confidence that the problem was identified and we could figure out how to prevent it going forward, but we are at this point unblocked on our network policy development.

Thanks so much for the help 🙏  If this does pop back up I'll re-open the issue so we can figure it out.
